src/vidpf.rs

```diff
@@ -0,0 +1,644 @@
+    const NODE_PROOF_ADJUST: &'static [u8] = b"NodeProofAdjust";
+}
+
+/// [`Key`] is the private key of an aggregation server.
+pub struct Key {
+    id: ServerId,
+    value: [u8; 16],
+}
+
+impl Key {
+    /// Generates a key at random.
+    ///
+    /// # Errors
+    /// Triggers an error if the random generator fails.
+    pub(crate) fn gen(id: ServerId) -> Result<Self, VidpfError> {
+        let mut value = [0; 16];
+        getrandom::getrandom(&mut value)?;
+        Ok(Self { id, value })
+    }
+}
+
+/// `ServerId` used to identify the two aggregation servers.
+#[derive(Clone, Copy, PartialEq, Eq)]
+pub(crate) enum ServerId {
+    /// S0 is the first server.
+    S0,
+    /// S1 is the second server.
+    S1,
+}
+
+impl From<ServerId> for Choice {
+    fn from(value: ServerId) -> Self {
+        match value {
+            ServerId::S0 => Self::from(0),
+            ServerId::S1 => Self::from(1),
+        }
+    }
+}
+
+/// [`CorrectionWord`] adjusts values of shares during the VIDPF evaluation.
+struct CorrectionWord<Weight: VidpfWeight> {
+    s_cw: Seed,
+    t_cw_l: Choice,
+    t_cw_r: Choice,
+    w_cw: Weight,
+}
+
+/// [`PublicShare`] is the common public information used by aggregation servers.
+pub struct PublicShare<Weight: VidpfWeight> {
+    cw: Vec<CorrectionWord<Weight>>,
+    cs: Vec<Proof>,
+}
+
+/// [`EvalState`] contains the values produced during input evaluation at a given level.
+pub struct EvalState {
+    level: usize,
+    seed: Seed,
+    t_control_bit: Choice,
+    proof: Proof,
+}
+
+impl EvalState {
+    fn init_from_key(key: &Key) -> Self {
+        Self {
+            level: 0,
+            seed: key.value,
+            t_control_bit: Choice::from(key.id),
+            proof: Proof::default(),
+        }
+    }
+}
+
+/// [`ValueShare`] contains a share of the input's weight together with a proof for verification.
+pub struct ValueShare<Weight: VidpfWeight> {
+    /// Secret share of the input's weight.
+    pub(crate) share: Weight,
+    /// Proof used to verify the share.
+    pub(crate) proof: Proof,
+}
+
+/// Proof size in bytes.
+const VIDPF_PROOF_SIZE: usize = 32;
+
+/// [`Proof`] is used to validate user input and shares after evaluation.
+#[derive(Debug, PartialEq, Eq, Default)]
+pub struct Proof([u8; VIDPF_PROOF_SIZE]);
+
+impl Proof {
+    fn xor(mut self, other: &Self) -> Self {
+        zip(&mut self.0, other.0).for_each(|(a, b)| a.bitxor_assign(b));
+        self
+    }
+
+    fn conditional_xor(mut self, choice: Choice, other: &Self) -> Self {
+        zip(&mut self.0, other.0).for_each(|(a, b)| a.conditional_assign(&a.bitxor(b), choice));
+        self
+    }
+}
+
+/// [`Seed`] is used to feed a pseudorandom generator during evaluation.
+type Seed = [u8; 16];
+
+/// [`Weight`] represents a vector of field elements that implements the [`VidpfWeight`] trait.
+#[derive(Debug, PartialEq, Eq)]
+pub struct Weight<F: FieldElement, const N: usize>([F; N]);
```

Comment:
I don't think we should have `N` as a const generic. Generally the length of the weight is known only at runtime because it's determined by the type of data we want to aggregate; making `N` generic forces us to choose this at compile time.

That said, I think it would be useful to have `impl<F: FieldElement> VidpfWeight` for `Weight<F>`. where
```rust
pub struct Weight<F>(Vec<F>);
```

Reply:
done